[WIP] 뱀 게임 만들기

junghyeonsu opened the pull request:
- Closes #11

junghyeonsu committed:
feat: SnakeGameCanvas 생성

junghyeonsu committed:
chore: 다른 색깔 찾기에서 뱀 게임으로 변경

junghyeonsu committed:
chore: components import 관리

junghyeonsu committed:
feat: 캔버스 오른쪽 이동까지 구현

junghyeonsu committed:
chore: next.config.js Prettier 적용

junghyeonsu committed:
chore: 커스텀 훅 미리 생성해놓기

- useCanvas
- useScore

junghyeonsu committed:
feat: 뱀 게임 초기 세팅 완료

- 이동 구현

junghyeonsu committed:
chore: timer 변수 ref로 변경

junghyeonsu committed:
chore: 타입 이름 "I" prefix 추가 및 상수 적용

junghyeonsu committed:
feat: 뱀 몸 추가하기 & 음식 위치 랜덤 구현

- 음식을 먹었을 때 뱀의 몸 길이를 1 늘림
- 음식을 먹었을 때 음식의 위치가 랜덤으로 바뀜

junghyeonsu committed:
feat: 길이 진행도 생성

- 타입 폴더 생성 및 정리
- components/SnakeGame 폴더 생성해서 그 안에 컴포넌트들 정리

junghyeonsu committed:
feat: 뱀이 바깥으로 이동했을 때 처리

- 게임 보드 밖을 나갔을 때 처리
- 현재는 alert로 처리

junghyeonsu committed:
chore: 뱀 관련 타입 폴더명 변경

- snakeGame -> snake

junghyeonsu committed:
feat: 뱀 몸 충돌 처리

- hasSnakeCollided snake 유틸 함수 생성

junghyeonsu committed:
feat: 반대 방향 이동 금지 구현

junghyeonsu committed:
feat: 먹이를 먹을수록 속도가 빨라지도록 구현

junghyeonsu committed:
chore: 매직 넘버 상수화

junghyeonsu committed:
chore: 구현 완료 TODO 주석 삭제

junghyeonsu committed:
chore: TODO 주석 추가

junghyeonsu committed:
chore: SnakeGameBodyCounter Progressbar 제거

junghyeonsu committed:
refactor: hasSnakeCollided util 함수 수정

- 중간에서 부딪혔으면 끝까지 확인하지 않고 부딪힌 것 확인하자마자 바로 리턴하도록 수정

junghyeonsu committed:
refactor: isSnakeOutOfCanvas 유틸함수 생성

- 뱀이 바깥으로 나갔는지 확인하는 함수

junghyeonsu committed:
chore: hasSnakeCollided 유틸 함수 이름 변경

- hasSnakeCollided에서 isSnakeCollided로 변경
- is prefix로 컨벤션 맞춤

junghyeonsu committed:
chore: 캔버스 관련 상수 생성

- 실제 width, height가 반영되지 않고, 조금 다른 숫자가 반영됨
- 왜 그런지는 모르겠음
- 우선 x, y 각각 min, max값을 상수로 선언

junghyeonsu committed:
chore: 뱀 게임 유틸 함수 매개변수명 변경

- 조금 더 직관적으로 변경

junghyeonsu committed:
refactor: isSnakeEatFood 함수 생성

- 뱀이 음식을 먹었는지 확인하는 유틸 함수

junghyeonsu committed:
feat: 사과로 food 대체

junghyeonsu committed:
chore: 뱀 게임 난이도 조절

- 제한 프레임 속도 30으로 변경

junghyeonsu committed:
chore: 주석 제거

junghyeonsu committed:
feat: 스피드 Progress bar 생성

junghyeonsu committed:
chore: 프레임 시작 값 상수 선언

junghyeonsu commented on src/components/SnakeGame/SnakeGameBodyCounter.tsx:
```diff
@@ -0,0 +1,10 @@
+import { memo } from 'react';
+import { Box } from '@chakra-ui/react';
+
+import type { ISnakeGameHook } from '../../types/snake';
+
+const SnakeGameBodyCounter = ({ snakeBodyLength }: Pick<ISnakeGameHook, 'snakeBodyLength'>) => {
+  return <Box>{snakeBodyLength}</Box>;
+};
+
+export default memo(SnakeGameBodyCounter);
```
![스크린샷 2022-05-12 오후 10 48 02](https://user-images.githubusercontent.com/54893898/168090205-417bbb26-964f-4299-bd5a-60dcaaec5bbc.png)

위에 뱀 몸통 길이가 얼마인지 나타내는 컴포넌트

junghyeonsu commented on src/components/SnakeGame/SnakeGameCanvas.tsx:
```diff
@@ -0,0 +1,93 @@
+      snakeGameDispatch({ type: SNAKE_ACTIONS.MOVE });
+    }
+
+    if (isSnakeEatFood(snakeBody[0], foodPosition)) {
+      if (currentFrame > SNAKE.FRAME_DOWN_LIMIT)
+        snakeGameDispatch({ type: SNAKE_ACTIONS.CHANGE_FRAME });
+      snakeGameDispatch({ type: SNAKE_ACTIONS.ADD_SNAKE_BODY });
+      snakeGameDispatch({ type: SNAKE_ACTIONS.CHANGE_FOOD_POSITION });
+    }
```
- 뱀이 음식을 먹었는지 확인하는 로직
- `currentFrame > SNAKE.FRAME_DOWN_LIMIT` 이 부분은 frame이 음식 먹을 때 마다 줄어들어서 점점 뱀이 빨라지는데 그 속도의 한계를 제한하기 위해서 프레임에 최소 한계를 정해놓고, 그 한계에 도달하면 더 이상 줄어들지 못하도록 설정

junghyeonsu commented on src/components/SnakeGame/SnakeGameCanvas.tsx:
```diff
@@ -0,0 +1,93 @@
+      snakeGameDispatch({ type: SNAKE_ACTIONS.CHANGE_FOOD_POSITION });
+    }
+
+    if (isSnakeOutOfCanvas(snakeBody[0])) {
+      snakeGameDispatch({ type: SNAKE_ACTIONS.RESET });
+      alert('뱀이 게임 밖으로 나갔습니다.');
+    }
```
- 뱀이 캔버스 바깥으로 나갔는지 확인하는 로직
- 나갔으면 alert해주고, 게임을 다시 리셋

junghyeonsu commented on src/components/SnakeGame/SnakeGameCanvas.tsx:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    // TODO: 충돌을 넣을까 말까? 고민해보기
+    if (isSnakeCollided(snakeBody, snakeBody[0])) {
+      snakeGameDispatch({ type: SNAKE_ACTIONS.RESET });
+      alert('뱀이 몸이랑 부딪혔습니다.');
+    }
```
- 뱀이 자신 몸에 충돌했는지 확인하는 로직
- 부딪히면 alert해주고 게임을 다시 리셋

junghyeonsu commented on src/components/SnakeGame/SnakeGameCanvas.tsx:
```diff
@@ -0,0 +1,93 @@
+
+    drawLine(context, SNAKE.CANVAS_WIDTH, SNAKE.CANVAS_HEIGHT);
+    drawObject(context, snakeBody, SNAKE.SNAKE_BODY_COLOR);
+    drawImage(context, foodPosition, appleImageRef.current);
```
매 프레임마다 캔버스에 있는 오브젝트들 (세로 가로 회색 줄들, 뱀, 사과 이미지)을 그려준다.

junghyeonsu commented on src/components/SnakeGame/SnakeGameCanvas.tsx:
```diff
@@ -0,0 +1,93 @@
+    requestAnimationRef.current = requestAnimationFrame(render);
+  }, [context, currentFrame, foodPosition, snakeBody, snakeGameDispatch]);
+
+  useEffect(() => {
+    const appleImage = new Image();
+    appleImage.src = '/snake/apple.png';
+    appleImageRef.current = appleImage;
+  }, []);
```
사과 이미지 맨 처음에 받아오기

junghyeonsu commented on src/components/SnakeGame/SnakeGameCanvas.tsx:
```diff
@@ -0,0 +1,93 @@
+    return () => cancelAnimationFrame(requestAnimationRef.current);
+  }, [render]);
+
+  useEffect(() => {
+    window.addEventListener('keydown', handleKeyDown);
+    return () => window.removeEventListener('keydown', handleKeyDown);
+  }, [handleKeyDown]);
```
맨 처음에 키보드 핸들러를 붙여줌

junghyeonsu commented on src/components/SnakeGame/SnakeSpeedProgressbar.tsx:
```diff
@@ -0,0 +1,31 @@
+import { memo } from 'react';
+import { Progress } from '@chakra-ui/react';
+
+import { SNAKE } from '../../constants';
+
+import type { ISnakeGameHook } from '../../types/snake';
+
+const SnakeSpeedProgressbar = ({ currentFrame }: Pick<ISnakeGameHook, 'currentFrame'>) => {
+  const currentSpeedPercentage =
+    ((100 - currentFrame) / (SNAKE.FRAME_START - SNAKE.FRAME_DOWN_LIMIT)) * 100 - 100;
+
+  const colorScheme = (currentSpeedPercentage: number) => {
+    if (currentSpeedPercentage < 20) return 'green';
+    if (currentSpeedPercentage < 40) return 'blue';
+    if (currentSpeedPercentage < 70) return 'orange';
+    return 'red';
+  };
+
+  return (
+    <Progress
+      colorScheme={colorScheme(currentSpeedPercentage)}
+      hasStripe
+      isAnimated
+      value={currentSpeedPercentage}
+      width="360px"
+      height="30px"
+    />
+  );
+};
+
+export default memo(SnakeSpeedProgressbar);
```
![스크린샷 2022-05-12 오후 10 48 02](https://user-images.githubusercontent.com/54893898/168091197-e3d5a0e5-8f1a-4e9c-8c3e-7be1341951ea.png)

- 사진에서 아래에 현재 스피드에 대한 퍼센트를 progress bar로 표현
- 20% 아래는 초록색, 40% 아래는 파란색, 70% 아래는 주황색, 100%까지는 빨강색으로 표현
- React.memo에 대해서 궁금하다면 [요기를](https://ui.toast.com/weekly-pick/ko_20190731) 참조

junghyeonsu commented on src/components/SnakeGame/index.tsx:
```diff
@@ -0,0 +1,42 @@
+import SnakeGameCanvas from './SnakeGameCanvas';
+import SnakeGameBodyCounter from './SnakeGameBodyCounter';
+import SnakeSpeedProgressbar from './SnakeSpeedProgressbar';
+
+import { useSnakeGame } from '../../hooks';
+
+const SnakeGame = () => {
+  const {
+    snakeBody,
+    snakeBodyLength,
+    foodPosition,
+    currentFrame,
+    snakeGameDispatch,
+    handleKeyDown,
+  } = useSnakeGame();
+
+  return (
+    <Box
+      width="100%"
+      height="100%"
+      display="flex"
+      flexDirection="column"
+      justifyContent="center"
+      alignItems="center"
+      rowGap="20px"
+    >
+      <SnakeGameBodyCounter snakeBodyLength={snakeBodyLength} />
+      <SnakeGameCanvas
+        snakeBody={snakeBody}
+        foodPosition={foodPosition}
+        snakeGameDispatch={snakeGameDispatch}
+        handleKeyDown={handleKeyDown}
+        currentFrame={currentFrame}
+      />
+      <SnakeSpeedProgressbar currentFrame={currentFrame} />
+    </Box>
+  );
+};
+
+export default SnakeGame;
```
- `SnakeGame` 게임 전체를 나타내는 컴포넌트
- 해당 컴포넌트는 `pages/game/snake.ts`에 들어가는 유일한 파일
- 요기서 useSnake 훅을 불러와서 모든 데이터를 아래 컴포넌트에 뿌려줌
- [container-presenter 패턴](https://velog.io/@dolarge/React-Container-Presenter-%ED%8C%A8%ED%84%B4)을 사용해봤음

bae-sh replied:
continer-presenter 패턴의 의미가 index.ts가 container를 담당하고 SnakeGame 컴포넌트들이 presenter의 역할을 하는거야??

junghyeonsu replied:
우리가 store 만들면 `Provider`를 사용하는데, `Provider`를 앱 전체에 감싸잖아?
그거를 나는 `SnakeGame/index.tsx` 파일에 Provider를 붙여서 쓰는 거다? 라고 생각하면 조금 이해가 가려나
ㅇㅇㅇㅇ `index` 파일이 컨테이너이고, 나머지 컴포넌트들은 컨테이너에서 데이터를 받아서 보여주기만 하는 형식

junghyeonsu commented on src/constants/index.ts:
```diff
@@ -0,0 +1,4 @@
+import ROUTES from './routes';
+import SNAKE from './snake';
+
+export { ROUTES, SNAKE };
```
상수 파일들의 Import를 관리하는 index 파일

junghyeonsu commented on src/constants/routes.ts:
```diff
@@ -0,0 +1,5 @@
+  DYNO: '/game/dyno',
+  SNAKE: '/game/snake',
+  MEMORY: '/game/memory',
+});
```
우리 앱의 모든 경로를 관리하는 상수 파일

junghyeonsu commented on src/constants/snake.ts:
```diff
@@ -0,0 +1,24 @@
+export default Object.freeze({
+  CANVAS_WIDTH: 360,
+  CANVAS_HEIGHT: 400,
+  CANVAS_REAL_WIDTH_MIN: 0,
+  CANVAS_REAL_HEIGHT_MIN: 0,
+  CANVAS_REAL_WIDTH_MAX: 290,
+  CANVAS_REAL_HEIGHT_MAX: 145,
+
+  SNAKE_BODY_WIDTH_UNIT: 10,
+  SNAKE_BODY_HEIGHT_UNIT: 5,
+  FRAME_DOWN_UNIT: 1,
+  FRAME_DOWN_LIMIT: 30,
+  FRAME_START: 65,
+  TICK_TOC_UNIT: 10,
+
+  SNAKE_BODY_COLOR: '#f5046d',
+  FOOD_COLOR: '#676FA3',
+
+  SNAKE_LEFT_DIRECTION: 'LEFT',
+  SNAKE_RIGHT_DIRECTION: 'RIGHT',
+  SNAKE_UP_DIRECTION: 'UP',
+  SNAKE_DOWN_DIRECTION: 'DOWN',
+  SNAKE_STOP: 'STOP',
+});
```
뱀 게임 관련 상수 값들 관리

bae-sh replied:
상수들도 사용 목적에 따라서 파일을 나누는 구나

junghyeonsu replied:
이것도 취향차이인데 우리 프로젝트가 상수가 커질 것 같아서 이렇게 했으

junghyeonsu commented on src/hooks/useSnakeGame.tsx:
```diff
@@ -0,0 +1,177 @@
+import { useCallback, useReducer } from 'react';
```
해당 파일은 `snake` 관련 리듀서, 액션이 있는 useSnakeGame 커스텀 훅

`useSnakeGame hook`에 있는 `dispatch`를 통해서 `action`을 발동
그러면 `Reducer`에서는 `Action`에 따라서 state들을 변경시켜주는 방식

junghyeonsu commented on src/utils/canvas.ts:
```diff
@@ -0,0 +1,74 @@
+
+import { SNAKE } from '../constants';
+
+export const drawLine = (
+  context: CanvasRenderingContext2D | null,
+  width: number,
+  height: number,
+) => {
+  if (context) {
+    Array.from({ length: width / SNAKE.SNAKE_BODY_WIDTH_UNIT }, () => 0).forEach((_, idx) => {
+      context.beginPath();
+      context.lineWidth = 0.1;
+      context.moveTo(idx * SNAKE.SNAKE_BODY_WIDTH_UNIT, 0);
+      context.lineTo(idx * SNAKE.SNAKE_BODY_WIDTH_UNIT, height);
+      context.strokeStyle = 'gray';
+      context.stroke();
+    });
+
+    Array.from({ length: height / SNAKE.SNAKE_BODY_HEIGHT_UNIT }, () => 0).forEach((_, idx) => {
+      context.beginPath();
+      context.lineWidth = 0.1;
+      context.moveTo(0, idx * SNAKE.SNAKE_BODY_HEIGHT_UNIT);
+      context.lineTo(width, idx * SNAKE.SNAKE_BODY_HEIGHT_UNIT);
+      context.strokeStyle = 'gray';
+      context.stroke();
+    });
+  }
+};
```
현재 캔버스에 가로, 세로 바둑판을 그려주는 캔버스 유틸 함수

junghyeonsu commented on src/utils/canvas.ts:
```diff
@@ -0,0 +1,74 @@
+  }
+};
+
+export const drawObject = (
+  context: CanvasRenderingContext2D | null,
+  objectBody: IObjectBody[],
+  fillColor: string,
+) => {
+  if (context) {
+    objectBody.forEach((object: IObjectBody) => {
+      context.fillStyle = fillColor;
+      context.strokeStyle = 'gray';
+      context.fillRect(
+        object.x,
+        object.y,
+        SNAKE.SNAKE_BODY_WIDTH_UNIT,
+        SNAKE.SNAKE_BODY_HEIGHT_UNIT,
+      );
+      context.strokeRect(
+        object.x,
+        object.y,
+        SNAKE.SNAKE_BODY_WIDTH_UNIT,
+        SNAKE.SNAKE_BODY_HEIGHT_UNIT,
+      );
+    });
+  }
+};
```
오브젝트를 그려주는 유틸 함수이긴 한데,,, 지금은 뱀 게임에 맞춰져있는 느낌?..

junghyeonsu commented on src/utils/canvas.ts:
```diff
@@ -0,0 +1,74 @@
+  }
+};
+
+export const drawImage = (
+  context: CanvasRenderingContext2D | null,
+  objectPosition: IObjectBody,
+  image: HTMLImageElement | null,
+): void => {
+  if (image) {
+    context?.drawImage(
+      image,
+      objectPosition.x,
+      objectPosition.y,
+      SNAKE.SNAKE_BODY_WIDTH_UNIT * 2,
+      SNAKE.SNAKE_BODY_HEIGHT_UNIT * 2,
+    );
+  }
+};
```
해당 유틸 함수도 이미지를 그려주는 유틸함수인데, 지금은 뱀 게임에 맞춰져서 로직이 짜여진듯.
그래서 나중에 꼭 바꾸도록 하겠습니다.

sumi-0011 replied:
제가 사용하는 drawImage는
```
export const drawImage = (ctx: CanvasRenderingContext2D | null, object: ICanvasObject) => {
  if (!ctx) return;
  const img: CanvasImageSource = object.image as HTMLImageElement;
  ctx.drawImage(img, object.x, object.y, object.width, object.height);
};
```
이여서 
```
 SNAKE.SNAKE_BODY_WIDTH_UNIT * 2,
      SNAKE.SNAKE_BODY_HEIGHT_UNIT * 2,
```
말 고는 큰 차이는 없는거 같아요! 이 부분만 수정하면 같은 함수로 사용할수있을것 같습니당

junghyeonsu replied:
오오 오키오키 한 번 유틸로 만들어서 우리 둘 다 사용할 수 있도록 한번 해볼게

junghyeonsu commented on src/utils/canvas.ts:
```diff
@@ -0,0 +1,74 @@
+  }
+};
+
+export const clearBoard = (context: CanvasRenderingContext2D | null) => {
+  if (context) context.clearRect(0, 0, SNAKE.CANVAS_WIDTH, SNAKE.CANVAS_HEIGHT);
+};
```
해당 유틸 함수도 캔버스를 지워주는 유틸함수인데, 지금은 뱀 게임에 맞춰져서 로직이 짜여진듯.
그래서 나중에 꼭 바꾸도록 하겠습니다.

sumi-0011 replied:
```

  const clearCanvas = () => {
    canvasRef.current &&
      context &&
      context.clearRect(0, 0, canvasRef.current.width, canvasRef.current.height);
  };
```
이렇게 사용하고 있어 이부분도 큰 차이는 없는 것 같습니다!

junghyeonsu commented on src/utils/snake.ts:
```diff
@@ -0,0 +1,32 @@
+import { SNAKE } from '../constants';
```
보통은 이 파일에 작성된 함수들이 `util`성을 가진다고 보기에는 조금 어려운 것 같음.
`util` 함수 같은건 앱 통틀어서 공통적으로 사용될 수 있는 함수 같은 느낌인데, 아래 함수들은 뱀 게임에서만 작동하는 그런 느낌.

그래서 `util`보다는 `service`에 조금 더 가까운 느낌인데 이 부분은 나도 어디다 둬야할 지 아직 잘 몰라서 우선은 `util`에 뒀어
혹시 의견들 있니?

sumi-0011 replied:
몇몇 게임 관련 메소드들을 한꺼번에 두지 않고 깔끔하게 관리하려고 그런거죠?!
저도 가능하면 이렇게 게임 관련 로직을 다른 파일에 분리 해두고 싶은데 
`util`도 애매하고 다른 폴더에 넣기도 그래서 아직은 분리하지 않고 있는 상태예요

bae-sh replied:
게임 관련 로직과 관련된 하나의 폴더를 생성하고 그곳에서 게임 별로 폴더 생성 후 관리해보면 어떨까 하는데.. 나도 로직 관련 함수를 따로 빼고 싶어서!

junghyeonsu replied:
아하 오키! 이 부분에 공통적인 관심사가 있네
그럼 service 라는 폴더를 만들어서 거기서 파일 나눠가지고 사용하면 되겠다. 일단은 오케이!
일단 이런 문제가 있다는 걸 알고 있을게 내가

junghyeonsu reviewed (commented):
![gif](https://user-images.githubusercontent.com/54893898/168095374-a1ffba4b-9be0-486f-8dd0-e15a45feb2a5.gif)

@bae-sh  @sumi-0011 

뱀 게임 우선 잘 동작하도록 개발 했고, 추가로 하고싶은 것들이 많은데 (콤보, 애니메이션 효과 등등) 근데 지금은 로직은 다 짜여졌으니 다른 것들 개발(헤더, 푸터, 메인 페이지, 로그인)을 해야하는 것도 많아서 우선 요기서 스탑하고 리뷰 올림!

아래에 리뷰들 다 적어놨으니, 참고하고 모르는 것이나 이상한 거 있으면 코멘트 달아주면 돼!

bae-sh commented on src/components/SnakeGame/SnakeGameBodyCounter.tsx:
```diff
@@ -0,0 +1,10 @@
+
+import type { ISnakeGameHook } from '../../types/snake';
+
+const SnakeGameBodyCounter = ({ snakeBodyLength }: Pick<ISnakeGameHook, 'snakeBodyLength'>) => {
```
Pick<ISnakeGameHook, 'snakeBodyLength'> 여기서 pick은 처음보는데 언제 사용하는거야??

sumi-0011 commented on src/components/SnakeGame/SnakeGameBodyCounter.tsx:
```diff
@@ -0,0 +1,10 @@
+
+import type { ISnakeGameHook } from '../../types/snake';
+
+const SnakeGameBodyCounter = ({ snakeBodyLength }: Pick<ISnakeGameHook, 'snakeBodyLength'>) => {
+  return <Box>{snakeBodyLength}</Box>;
+};
```
`Pick<ISnakeGameHook, 'snakeBodyLength'>`에서
Pick이 뭔지 알수있을까요??!

sumi-0011 reviewed (approved):
확인했습니다 😄 
공통적으로 사용하는 함수들만 고쳐서 같이 사용하면 될것 같아요!

bae-sh commented on src/types/snake.ts:
```diff
@@ -0,0 +1,37 @@
+  snakeGameDispatch: React.Dispatch<SnakeGameAction>;
+  // eslint-disable-next-line no-unused-vars
+  handleKeyDown: (event: KeyboardEvent) => void;
+}
```
여기서 extends의 역할이 뭐야?? 내가 아는 상속 개념이랑 비슷한건가??

bae-sh commented on src/components/SnakeGame/SnakeSpeedProgressbar.tsx:
```diff
@@ -0,0 +1,31 @@
+import type { ISnakeGameHook } from '../../types/snake';
+
+const SnakeSpeedProgressbar = ({ currentFrame }: Pick<ISnakeGameHook, 'currentFrame'>) => {
+  const currentSpeedPercentage =
+    ((100 - currentFrame) / (SNAKE.FRAME_START - SNAKE.FRAME_DOWN_LIMIT)) * 100 - 100;
```
currentSpeedPercentage를 state가 아닌 상수로 관리를 하는 이유가 있을까??

bae-sh commented on src/utils/snake.ts:
```diff
@@ -0,0 +1,32 @@
+  SnakeBodyPosition: IObjectBody[],
+  SnakeHeadPosition: IObjectBody,
+): boolean => {
+  for (let i = 0; i < SnakeBodyPosition.length; i++) {
+    const isCurrentCollided =
+      SnakeBodyPosition[i].x === SnakeHeadPosition.x &&
+      SnakeBodyPosition[i].y === SnakeHeadPosition.y &&
+      i !== 0;
+    if (isCurrentCollided) return true;
+  }
```
```suggestion
    SnakeBodyPosition.forEach(({ x, y }, index) => {
    const isCurrentCollided = 
		x === SnakeHeadPosition.x && 
		y === SnakeHeadPosition.y && 
		index !== 0;

    if (isCurrentCollided) return true;
  });
```

위 for문을 좀 더 js 스럽게 forEach로 사용해보면 어때??

junghyeonsu replied:
이게 for문이랑 forEach문이 다른 점이
`for`문은 어떤 조건이 만족될 때 바로 return 할 수 있는 return을 제공하는데,
`forEach`안에서는 return문을 넣어도 바로 `forEach`가 종료되지 않아.
그리고 무조건 true를 반환하게 되어있어서 원하는대로 동작을 안할거야. 그래서 for문을 사용했으!

bae-sh replied:
오오 그런차이가 있었구나!! 하나또 배워갑니다 😄

bae-sh commented on src/hooks/useSnakeGame.tsx:
```diff
@@ -0,0 +1,177 @@
+  CHANGE_FRAME: 'CHANGE_FRAME',
+};
+
+const SNAKE_BODY_INITIAL = [
+  { x: SNAKE.SNAKE_BODY_WIDTH_UNIT * 9, y: SNAKE.SNAKE_BODY_HEIGHT_UNIT * 5 },
+  { x: SNAKE.SNAKE_BODY_WIDTH_UNIT * 8, y: SNAKE.SNAKE_BODY_HEIGHT_UNIT * 5 },
+  { x: SNAKE.SNAKE_BODY_WIDTH_UNIT * 7, y: SNAKE.SNAKE_BODY_HEIGHT_UNIT * 5 },
+  { x: SNAKE.SNAKE_BODY_WIDTH_UNIT * 6, y: SNAKE.SNAKE_BODY_HEIGHT_UNIT * 5 },
+  { x: SNAKE.SNAKE_BODY_WIDTH_UNIT * 5, y: SNAKE.SNAKE_BODY_HEIGHT_UNIT * 5 },
+];
+
+const INITIAL_STATE: ISnakeState = {
+  snakeBody: SNAKE_BODY_INITIAL,
+  snakeDirection: SNAKE.SNAKE_RIGHT_DIRECTION,
+  snakeBodyLength: 5,
+  foodPosition: {
+    x: 150,
+    y: 75,
+  },
+  currentFrame: SNAKE.FRAME_START,
+};
```
요부분은 상수 파일에 두는건 어때??

junghyeonsu replied:
이거를 진짜 많이 고민했는데, 저게 "초기값" 인데 보통 초기값은 리듀서랑 같이두는게 맞는 것 같아서 이렇게 두긴 했는데
그리고 저 값들이 상수로 써놓긴 했는데 상수보단 "초기값" 이라는 의미에 조금 더 취중되는 값인 것 같아서 저렇게 뒀으!
네가 말한 것도 한 번 고려를 해봐야겠다..

bae-sh commented on src/hooks/useSnakeGame.tsx:
```diff
@@ -0,0 +1,177 @@
+};
+
+const useSnakeGame = (): ISnakeGameHook => {
+  const [
+    { snakeBody, snakeDirection, foodPosition, snakeBodyLength, currentFrame },
+    snakeGameDispatch,
+  ] = useReducer(reducer, INITIAL_STATE);
```
useReduer기 하는 역할이 뭐야??

junghyeonsu replied:
useState 대신에 useReducer를 쓸 수 있는데 보통은 reducer를 만들어내는 훅스야.

useState 대신에 useReducer를 써서 얻을 수 있는 이점은
- 예측 가능하고 일관되게 작동하며, 그래서 복잡한 상태를 관리하는 데 적합합니다.
- 다음 state가 이전 state에 의존적인 경우
- 일반적으로 테스트하기 쉽습니다.
- 실행 취소/다시 실행, 상태 지속성 등과 같은 헬퍼 함수는 reducer(리듀셔)로 구현하기가 더 쉽습니다.
[참고](https://cpro95.tistory.com/642?category=929244)

위와 같고, 보통 관리해야 하는 상태가 많을 때는 useState를 사용해서 관리하는 것 보다,
`useReducer`를 사용해서 관리를 하는 것 같아.

bae-sh replied:
오오 나도 기회되면 `useReducer` 를 사용해봐야겠군!

junghyeonsu committed:
chore: snake 관련 로직들 services 폴더로 이동

junghyeonsu committed:
refactor: 코드를 깔끔하게 수정

- styles/colors 파일만들어서 색 관리
- services 폴더 만들어서 각 로직들 관리
- useSnakeGameHook에서 있던 키보드 이벤트 핸들러를 SnakeGameCanvas로 이동

junghyeonsu committed:
docs: 리드미 수정

junghyeonsu committed:
refactor: reducer payload 타입 생성

junghyeonsu committed:
chore: HOME route 상수 추가
